### PR TITLE
fix compilation under MINGW

### PR DESCRIPTION
**Pull request by merlokk**
## Description
Under **MINGW** there is no `gettimeofday`. use `mingw_gettimeofday` instead of `gettimeofday` 

## Status
**READY**

## Requires Backporting
may be backported

## Migrations
**NO**

## Steps to test or reproduce
compile under **MINGW**

### CODE REVIEW

**Comment by RonEld** [comment]
@merlokk Thank you for your contribution!

In addition, unfortunately our policy is to not accept contributions, without a Contributor’s Licence Agreement (CLA) signed or authorised by yourself or your employer.

If this is a personal contribution, the easiest way to do this is if you create an mbed account and [accept this click through agreement](https://os.mbed.com/contributor_agreement/). Alternatively, you can find a [slightly different agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us, and is applicable if you don't want to create an mbed account or alternatively if this is a corporate contribution.

Thanks for your understanding and again, thanks for the contribution!

**Comment by merlokk** [comment]
Accepted:
username olegmsn
You accepted the Contributor Agreement on Thu 08 Nov 2018.

**Comment by RonEld** [comment]
@merlokk Thank you for accepting the CLA!
Since github and Mbed are not linked, we have to manually search for user names.
I apologize for confusion, and thank you for your confirmation!

**Comment by merlokk** [comment]
Thank you)

**Comment by AndrzejKurek** [comment]
Hello @merlokk and thanks for your contribution!
Could you please describe the environment you're using to reproduce this problem?
System architecture, mingw version, the configuration file used (if different from default) and any other deviations from the standard setup?
I've tested it using mingw32 and mingw64 on a 64-bit Windows 10, using the default configuration, and haven't run into the problem you've described - the implementation starting in line 100 is used.

**Comment by merlokk** [comment]
I have fully updated MSYS2 with MINGW32

Windows 10 with Intel i5 CPU

config here
https://github.com/RfidResearchGroup/proxmark3/blob/master/common/mbedtls/config.h

```
mingw-w64-i686-assimp 4.1.0-2
mingw-w64-i686-binutils 2.30-5
mingw-w64-i686-bzip2 1.0.6-6
mingw-w64-i686-ca-certificates 20170211-2
mingw-w64-i686-crt-git 7.0.0.5245.edf66197-1
mingw-w64-i686-dbus 1.12.10-1
mingw-w64-i686-expat 2.2.6-1
mingw-w64-i686-fontconfig 2.13.1-1
mingw-w64-i686-freeglut 3.0.0-4
mingw-w64-i686-freetype 2.9.1-1
mingw-w64-i686-gcc 7.3.0-2
mingw-w64-i686-gcc-libs 7.3.0-2
mingw-w64-i686-gettext 0.19.8.1-6
mingw-w64-i686-giflib 5.1.4-2
mingw-w64-i686-glib2 2.58.1-1
mingw-w64-i686-gmp 6.1.2-1
mingw-w64-i686-graphite2 1.3.12-1
mingw-w64-i686-harfbuzz 2.0.2-1
mingw-w64-i686-headers-git 7.0.0.5245.edf66197-1
mingw-w64-i686-icu 62.1-1
mingw-w64-i686-icu-debug-libs 62.1-1
mingw-w64-i686-isl 0.19-1
mingw-w64-i686-jasper 2.0.14-1
mingw-w64-i686-lcms2 2.9-1
mingw-w64-i686-libffi 3.2.1-4
mingw-w64-i686-libgcrypt 1.8.3-1
mingw-w64-i686-libgpg-error 1.32-1
mingw-w64-i686-libiconv 1.15-3
mingw-w64-i686-libxml2 2.9.8-1
mingw-w64-i686-libxslt 1.1.32-1
mingw-w64-i686-minizip 2.5.2-2
mingw-w64-i686-mpc 1.1.0-1
mingw-w64-i686-mpdecimal 2.4.2-1
mingw-w64-i686-mpfr 4.0.1-2
mingw-w64-i686-ncurses 6.1.20180908-1
mingw-w64-i686-openal 1.19.0-1
mingw-w64-i686-openssl 1.1.1-4
mingw-w64-i686-p11-kit 0.23.14-1
mingw-w64-i686-pcre 8.42-1
mingw-w64-i686-pcre2 10.32-1
mingw-w64-i686-python3 3.7.1-2
mingw-w64-i686-qt5 5.11.2-3
mingw-w64-i686-qtbinpatcher 2.2.0-2
mingw-w64-i686-readline 7.0.005-1
mingw-w64-i686-sqlite3 3.25.2-1
mingw-w64-i686-tcl 8.6.8-1
mingw-w64-i686-termcap 1.3.1-3
mingw-w64-i686-tk 8.6.8-1
mingw-w64-i686-vulkan 1.1.73.0-1
mingw-w64-i686-windows-default-manifest 6.4-3
mingw-w64-i686-wineditline 2.205-1
mingw-w64-i686-winpthreads-git 7.0.0.5247.01fe4924-1
mingw-w64-i686-xpm-nox 4.2.0-5
mingw-w64-i686-xz 5.2.4-1
mingw-w64-i686-z3 4.7.1-2
mingw-w64-i686-zlib 1.2.11-4
mingw-w64-i686-zziplib 0.13.69-1
```

**Comment by merlokk** [comment]
I think this line is different:
`//#define MBEDTLS_HAVE_ASM`

**Comment by merlokk** [comment]
done

**Comment by AndrzejKurek** [comment]
Sorry for the misleading commit earlier saying that you should append your description to an already-closed release version.
Since we just had a release, please add the following at the top:
`= mbed TLS x.x.x branch released xxxx-xx-xx
 Bugfix
   * Your description`
Thanks!

**Comment by merlokk** [comment]
done

**Comment by AndrzejKurek** [comment]
We're almost there! Just two small suggestions:
- Please add the "Bugfix" category above your description;
- Please use the sentence "Contributed by merlokk", since your nickname itself won't be understandable by people not knowing it :)
Thanks again!

File: ChangeLog [hunk 1]
```diff
@@ -1,5 +1,10 @@
 mbed TLS ChangeLog (Sorted per branch, date)
 
+= mbed TLS x.x.x branch released xxxx-xx-xx
+
```
Comment (hanno-arm) [inline ChangeLog]: This is lacking a section such as `Bugfix`.

File: ChangeLog [hunk 2]
```diff
@@ -1,5 +1,10 @@
...
+   * Fix compilation error on MINGW32/MINGW64 with deactivated option
+     `MBEDTLS_HAVE_ASM` in config file (@merlokk).
```
Comment (hanno-arm) [inline ChangeLog]: Minor: We don't add ticks around identifiers in the ChangeLog.

Comment (merlokk) [reply ChangeLog]: done

File: ChangeLog (hunk 1 above)
Comment (merlokk) [reply ChangeLog]: done

File: library/timing.c [hunk 3]
```diff
@@ -230,11 +230,19 @@ unsigned long mbedtls_timing_hardclock( void )
...
         gettimeofday( &tv_init, NULL );
+#endif
         hardclock_init = 1;
     }
 
+#ifdef __MINGW32__
+    mingw_gettimeofday( &tv_cur, NULL );
+#else
     gettimeofday( &tv_cur, NULL );
+#endif
```
Comment (hanno-arm) [inline library/timing.c]: Please add a `/* __MINGW32__ /*` comment here.

File: library/timing.c [hunk 4]
```diff
@@ -230,11 +230,19 @@ unsigned long mbedtls_timing_hardclock( void )
 
     if( hardclock_init == 0 )
     {
+#ifdef __MINGW32__
+        mingw_gettimeofday( &tv_init, NULL );
+#else
         gettimeofday( &tv_init, NULL );
+#endif
```
Comment (hanno-arm) [inline library/timing.c]: Please add a `/* __MINGW32__ /*` comment here.

**Comment by merlokk** [comment]
done

File: library/timing.c [hunk 5]
```diff
@@ -230,11 +230,19 @@ unsigned long mbedtls_timing_hardclock( void )
 
     if( hardclock_init == 0 )
     {
+#ifdef __MINGW32__
```
Comment (hanno-arm) [inline library/timing.c]: The ChangeLog entry says you're also fixing the case of mingw64. According to https://sourceforge.net/p/predef/wiki/Compilers/ `__MINGW64__`  implies `__MINGW32__`, so your code is fine, but if that's the underlying reasoning, I think we should add a comment, or simply add a clause for `__MINGW64__`.

Comment (hanno-arm) [inline library/timing.c]: Could you switch to `#if defined(XXX)` instead of `#ifdef XXX` please?

Comment (merlokk) [reply library/timing.c]: added

**Comment by merlokk** [comment]
done

File: library/timing.c [hunk 6]
```diff
@@ -230,11 +230,19 @@ unsigned long mbedtls_timing_hardclock( void )
...
         gettimeofday( &tv_init, NULL );
+#endif /* __MINGW32__ || __MINGW64__*/
         hardclock_init = 1;
     }
 
+#if defined(__MINGW32__) || defined(__MINGW64__)
+    mingw_gettimeofday( &tv_cur, NULL );
+#else
     gettimeofday( &tv_cur, NULL );
+#endif /* __MINGW32__ || __MINGW64__*/
```
Comment (hanno-arm) [inline library/timing.c]: Whitespace missing before `*/`.

**Comment by sbutcher-arm** [comment]
CI is failing on tabs in the ChangeLog, and the BSD timing test which is flakey.

@AndrzejKurek - Can you do a quick additional commit to remove the tabs? We can then get this merged. I don't need re-reviews, just for a silly fix in the ChangeLog.

**Comment by AndrzejKurek** [comment]
@sbutcher-arm - done.

**Comment by sbutcher-arm** [comment]
Now passing CI. Approved for merge. This PR should be merged to both `development` and `mbedtls-2.16`.

**Comment by sbutcher-arm** [comment]
I just approved this for merge, and then realised the previous approvals had been dismissed. I guess that was done automatically by @AndrzejKurek's new commit.

I didn't intend to require further reviews, but to help future auditing, I want approvals on the record, so @hanno-arm / @AndrzejKurek - can you please approve again?

**Comment by RonEld** [comment]
Backport to `mbedtls-2.16` available has been raised in #2507

**Comment by RonEld** [comment]
THis PR and its backport to `mbedtls-2.16` have been fully reviewed and approved, therefore removing the "needs backports" label.

cc @Patater

**Comment by sbutcher-arm** [comment]
The CI is failing on multiple counts. Some of these seem like known issues - DTLS 3d, BSD timing test etc - but ChangeLog is also failing for trailing whitespace.

Demoting to 'needs: work' - @RonEld - can you please fix the ChangeLog and ensure this passes CI?

**Comment by merlokk** [comment]
I can't look into CI( and I dont see any spaces in changelog(

**Pull request closed without merging**